`AvlTree.insert` operation

kettlebell opened the pull request:
Close #411

kettlebell commented on ergotree-interpreter/src/eval/savltree.rs:
```diff
@@ -0,0 +1,231 @@
+        &starting_digest,
+        &proof,
+        AVLTree::new(
+            |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
```
The Rust AVL library has this notion of `Resolver` that's not present in the scala implementation. I just copied what was used in the library example.

kettlebell commented on ergotree-interpreter/src/eval/savltree.rs:
```diff
@@ -0,0 +1,231 @@
+            AVLTree::new(
+                |digest| Node::LabelOnly(NodeHeader::new(Some(*digest), None)),
+                1,
+                None,
```
I'm not sure how this parameter is used for AVL operations. Setting it to None here and for `AvlTreeData::value_length_opt` allowed the test to pass.

greenhat commented:
`scorex_crypto_avltree` addition broke `ergo-lib-wasm` build by pulling new `rand`. Please update `rand` version in all our crates to `0.8.4`(or .3) and 
update `getrandom = {version = "0.1", features = ["wasm-bindgen"]}` in `ergo-lib-wasm` to `getrandom = {version = "0.2.3", features = ["js"]}`.
`

greenhat reviewed (changes requested):
Great job figuring out AVL trees! Looking great! Check minor comments.

greenhat commented:
> > `scorex_crypto_avltree` addition broke `ergo-lib-wasm` build by pulling new `rand`. Please update `rand` version in all our crates to `0.8.4`(or .3) and update `getrandom = {version = "0.1", features = ["wasm-bindgen"]}` in `ergo-lib-wasm` to `getrandom = {version = "0.2.3", features = ["js"]}`. `
> 
> I've updated the dependencies in my latest commit, but CI is totally broken as you can see. Each test refers to errors in ergotree-ir/src/mir/constant.rs, but the code doesn't match the commit. Not sure what's going on.
> 
> Also please note that I had to update more than just `rand` and `getrandom`. Details in commit message.

Judging by compiler error I believe you need to rebase this branch onto fresh develop, after I merged SubstConstants PR. For some reason Github UI did not pick up new develop branch.

kettlebell commented on ergotree-ir/src/mir/constant.rs:
```diff
@@ -53,6 +54,8 @@ pub enum Literal {
     /// Ergo box ID
     CBox(IrBoxId),
+    /// AVL tree
+    AvlTree(AvlTreeData),
```
Done.

kettlebell commented on ergotree-ir/src/mir/value.rs:
```diff
@@ -150,7 +151,7 @@ pub enum Value {
     CBox(IrBoxId),
     /// AVL tree
-    AvlTree,
+    AvlTree(AvlTreeData),
```
Done.

kettlebell commented on ergotree-ir/src/types/savltree.rs:
```diff
@@ -0,0 +1,53 @@
+    static ref INSERT_METHOD_DESC: SMethodDesc = SMethodDesc {
+        method_id: INSERT_METHOD_ID,
+        name: "insert",
+        tpe: SFunc {
```
Thanks.

kettlebell commented on ergotree-ir/src/mir/avl_tree_data.rs:
```diff
@@ -0,0 +1,135 @@
+        self.digest.sigma_serialize(w)?;
+        w.put_u8(self.tree_flags.0)?;
+        w.put_u32(self.key_length)?;
+        if let Some(l) = self.value_length_opt {
```
There's existing impl for `Option<Box<T>>` so I went with that (trying to implement it for `Option<T>` lead to compiler complaints).

kettlebell commented on ergotree-ir/src/mir/avl_tree_data.rs:
```diff
@@ -0,0 +1,135 @@
+        let digest = ADDigest::sigma_parse(r)?;
+        let tree_flags = AvlTreeFlags::parse(r.get_u8()?);
+        let key_length = r.get_u32()?;
+        let is_some = {
```
Done.

kettlebell commented on ergotree-interpreter/src/eval/savltree.rs:
```diff
@@ -0,0 +1,231 @@
+use super::EvalFn;
+
+pub(crate) static INSERT_EVAL_FN: EvalFn = |_env, _ctx, obj, args| {
+    let mut avl_tree_data = match obj {
```
Done.

kettlebell commented on ergotree-interpreter/src/eval/savltree.rs:
```diff
@@ -0,0 +1,231 @@
+    }?;
+
+    if !avl_tree_data.tree_flags.insert_allowed() {
+        return Err(EvalError::Misc("AvlTree: Insertions not allowed".into()));
```
I went with `EvalError::AvlTree(String)`.

kettlebell commented on ergotree-interpreter/src/eval/savltree.rs:
```diff
@@ -0,0 +1,231 @@
+            .get(0)
+            .cloned()
+            .ok_or_else(|| EvalError::NotFound("AvlTree: eval is missing first arg".to_string()))?;
+        match v {
```
Thanks, saved a lot of lines. Need to remember to use this trait more!

kettlebell commented on ergotree-ir/src/serialization/data.rs:
```diff
@@ -37,6 +37,7 @@ impl DataSerializer {
             Literal::GroupElement(ecp) => ecp.sigma_serialize(w)?,
             Literal::SigmaProp(s) => s.value().sigma_serialize(w)?,
             Literal::CBox(_) => return Err(SigmaSerializationError::NotImplementedYet("Box")),
+            Literal::AvlTree(a) => a.sigma_serialize(w)?,
```
Done.

kettlebell commented:
> > > `scorex_crypto_avltree` addition broke `ergo-lib-wasm` build by pulling new `rand`. Please update `rand` version in all our crates to `0.8.4`(or .3) and update `getrandom = {version = "0.1", features = ["wasm-bindgen"]}` in `ergo-lib-wasm` to `getrandom = {version = "0.2.3", features = ["js"]}`. `
> > 
> > 
> > I've updated the dependencies in my latest commit, but CI is totally broken as you can see. Each test refers to errors in ergotree-ir/src/mir/constant.rs, but the code doesn't match the commit. Not sure what's going on.
> > Also please note that I had to update more than just `rand` and `getrandom`. Details in commit message.
> 
> Judging by compiler error I believe you need to rebase this branch onto fresh develop, after I merged SubstConstants PR. For some reason Github UI did not pick up new develop branch.

Done, plus addressed all your concerns (with one little exception for the unresolved conversation above).

greenhat reviewed (approved):
Looking great! Thank you!
